rudder-core/src/main/scala/com/normation/rudder/domain/nodes/NodeInfo.scala

```diff
@@ -93,4 +98,100 @@ final case class NodeInfo(
   val properties                 = node.properties
   val policyMode                 = node.policyMode
 
+  lazy val cfengineKeyHash: String = {
+    publicKey.map(CFEngineKey.getHash) match {
+      case None =>
+        logger.info(s"Node '${hostname}' (${id.value}) doens't have a register public key")
+        ""
+      case Some(Full(hash)) =>
+        hash
+      case Some(eb:EmptyBox) =>
+        val e = eb ?~! s"Error when trying to get CFEngine public key hash of node '${hostname}' (${id.value})"
+        logger.error(e.messageChain)
+        ""
+    }
+  }
 }
+
+/*
+ * An object to deal with the specificities of CFEngine keys.
+ */
+object CFEngineKey {
+
+  /*
+   * CFengine public keys are store on the server in
+   * /var/rudder/cfengine-community/ppkeys/node-MD5-hash.pub .
+   *
+   * The content of these file is an RSA encoded key, something
+   * looking like:
+   *
+   *   -----BEGIN RSA PUBLIC KEY-----
+   *   MIIBCAKCAQEAv76gYG9OaFpc0eBeUXDM3WsRWyuHco3DpWnKrrpqQwylpEH26gRb
+   *   cu/L5KWc1ihj1Rv/AU3dkQL5KdXatSrWOLUMmYcQc5DYSnZacbdHIGLn11w1PHsw
+   *   9P2pivwQyIF3k4zqANtlZ3iZN4AXZpURI4VVhiBYPwZ4XgHPibcuJHiyNyymiHpT
+   *   HX9H0iaEIwyJMPjzRH+piFRmSeUylHfQLqb6AkD3Dg3Nxe9pbxNbk1saqgHFF4kd
+   *   Yh3O5rVto12XqisGWIbsmsT0XFr6V9+/sde/lpjI4AEcHR8oFYX5JP9/SXPuRJfQ
+   *   lEl8vn5PHTY0mMrNAcM7+rzpkOW2c7b8bwIBIw==
+   *   -----END RSA PUBLIC KEY-----
+   *
+   * You can see details with the following command:
+   *
+   * openssl rsa -RSAPublicKey_in -in /var/rudder/cfengine-community/ppkeys/node-MD5-hash.pub -text
+   *
+   * The hash by itself is the MD5 sum of the byte array
+   * of the modulus concatenated with the byte array of
+   * the exponent, both taken in big endian binary form.
+   * The corresponding openssl C code used in CFEngine for
+   * that is:
+   *   parse the key then
+   *   EVP_DigestInit(&context, md); # extract modulus in big endian binary form
+   *   actlen = BN_bn2bin(key->n, buffer);
+   *   EVP_DigestUpdate(&context, buffer, actlen); # extract exponent in big endian binary form
+   *   actlen = BN_bn2bin(key->e, buffer);
+   *   EVP_DigestUpdate(&context, buffer, actlen);
+   *   EVP_DigestFinal(&context, digest, &md_len);
+   *
+   * In Scala, we need to use bouncy castle, because nothing
+   * native read PEM / RSA file, and I don't like parsing
+   * ASN.1 that much.
+   *
+   * Oh, and Java Security API force a lot of casting.
+   * This is insane.
+   */
+  def getHash(key: PublicKey): Box[String] = {
+
+    //the parser able to read PEM files
+    val parser = new PEMParser(new StringReader(key.value))
+
```

Comment:
There are a lot of asInstanceOf - shouldn't you wrap it around a try to catch casting exception ?